Confusing Error message with EXPECT()

Mocking a method using `EXPECT` gives a confusing error messages when the call count for the method does not match the number of occurrences of this message in the code but the method signatures match. 

A more descriptive error message would be most helpful.

##### Output

```
└──➤ go test lib_test.go
--- FAIL: TestMyThing (0.00s)
	controller.go:129: no matching expected call: *mock_mypackage.MockMyInterface.SomeMethod([1 second])
FAIL
FAIL	command-line-arguments	0.006s
```
##### lib.go

```
package mypackage

type MyInterface interface {
	SomeMethod(x int, y string)
}
```

##### lib_test.go

```
package mypackage

import (
	"github.com/golang/mock/gomock"
	"github.com/jtestard/devIssues/lib_mock"
	"testing"
)

func TestMyThing(t *testing.T) {
	mockCtrl := gomock.NewController(t)
	defer mockCtrl.Finish()

	mockObj := mock_mypackage.NewMockMyInterface(mockCtrl)

	mockObj.EXPECT().SomeMethod(1, "second")

	mockObj.SomeMethod(1, "second")
	mockObj.SomeMethod(1, "second")
}
```

##### mockgen

```
mockgen -source=lib.go > lib_mock/lib_mock.go
```

The error message with the #129 would be:
--- FAIL: TestMyThing (0.00s)
        controller.go:148: Unexpected call to *lib_mock.MockMyInterface.SomeMethod([1 second]) at /home/ubuntu/workspace/src/test/lib_test.go:18 because: 
                Expected call at /home/ubuntu/workspace/src/test/lib_test.go:15 has already been called the max number of times.
FAIL

--- a/gomock/call.go
+++ b/gomock/call.go
@@ -66,7 +66,16 @@ func newCall(t TestReporter, receiver interface{}, method string, methodType ref
 	}
 
 	origin := callerInfo(3)
-	return &Call{t: t, receiver: receiver, method: method, methodType: methodType, args: margs, origin: origin, minCalls: 1, maxCalls: 1}
+	actions := []func([]interface{}) []interface{}{func([]interface{}) []interface{} {
+		// Synthesize the zero value for each of the return args' types.
+		rets := make([]interface{}, methodType.NumOut())
+		for i := 0; i < methodType.NumOut(); i++ {
+			rets[i] = reflect.Zero(methodType.Out(i)).Interface()
+		}
+		return rets
+	}}
+	return &Call{t: t, receiver: receiver, method: method, methodType: methodType,
+		args: margs, origin: origin, minCalls: 1, maxCalls: 1, actions: actions}
 }
 
 // AnyTimes allows the expectation to be called 0 or more times
@@ -390,18 +399,6 @@ func (c *Call) dropPrereqs() (preReqs []*Call) {
 	preReqs = c.preReqs
 	c.preReqs = nil
 	return
-}
-
-func (c *Call) defaultActions() []func([]interface{}) []interface{} {
-	return []func([]interface{}) []interface{}{func([]interface{}) []interface{} {
-		// Synthesize the zero value for each of the return args' types.
-		mt := c.methodType
-		rets := make([]interface{}, mt.NumOut())
-		for i := 0; i < mt.NumOut(); i++ {
-			rets[i] = reflect.Zero(mt.Out(i)).Interface()
-		}
-		return rets
-	}}
 }
 
 func (c *Call) call(args []interface{}) []func([]interface{}) []interface{} {

--- a/gomock/call.go
+++ b/gomock/call.go
@@ -44,6 +44,31 @@ type Call struct {
 	setArgs map[int]reflect.Value
 }
 
+// newCall creates a *Call. It requires the method type in order to support
+// unexported methods.
+func newCall(t TestReporter, receiver interface{}, method string, methodType reflect.Type, args ...interface{}) *Call {
+	if h, ok := t.(testHelper); ok {
+		h.Helper()
+	}
+
+	// TODO: check arity, types.
+	margs := make([]Matcher, len(args))
+	for i, arg := range args {
+		if m, ok := arg.(Matcher); ok {
+			margs[i] = m
+		} else if arg == nil {
+			// Handle nil specially so that passing a nil interface value
+			// will match the typed nils of concrete args.
+			margs[i] = Nil()
+		} else {
+			margs[i] = Eq(arg)
+		}
+	}
+
+	origin := callerInfo(3)
+	return &Call{t: t, receiver: receiver, method: method, methodType: methodType, args: margs, origin: origin, minCalls: 1, maxCalls: 1}
+}
+
 // AnyTimes allows the expectation to be called 0 or more times
 func (c *Call) AnyTimes() *Call {
 	c.minCalls, c.maxCalls = 0, 1e8 // close enough to infinity
@@ -283,6 +308,11 @@ func (c *Call) matches(args []interface{}) error {
 			return fmt.Errorf("Expected call at %s doesn't have a prerequisite call satisfied:\n%v\nshould be called before:\n%v",
 				c.origin, preReqCall, c)
 		}
+	}
+
+	// Check that the call is not exhausted.
+	if c.exhausted() {
+		return fmt.Errorf("Expected call at %s has already been called the max number of times.", c.origin)
 	}
 
 	return nil

--- a/gomock/callset.go
+++ b/gomock/callset.go
@@ -16,35 +16,47 @@ package gomock
 
 import (
 	"bytes"
-	"errors"
 	"fmt"
 )
 
 // callSet represents a set of expected calls, indexed by receiver and method
 // name.
-type callSet map[interface{}]map[string][]*Call
+type callSet struct {
+	// Calls that are still expected.
+	expected map[callSetKey][]*Call
+	// Calls that have been exhausted.
+	exhausted map[callSetKey][]*Call
+}
+
+// callSetKey is the key in the maps in callSet
+type callSetKey struct {
+	receiver interface{}
+	fname    string
+}
+
+func newCallSet() *callSet {
+	return &callSet{make(map[callSetKey][]*Call), make(map[callSetKey][]*Call)}
+}
 
 // Add adds a new expected call.
 func (cs callSet) Add(call *Call) {
-	methodMap, ok := cs[call.receiver]
-	if !ok {
-		methodMap = make(map[string][]*Call)
-		cs[call.receiver] = methodMap
+	key := callSetKey{call.receiver, call.method}
+	m := cs.expected
+	if call.exhausted() {
+		m = cs.exhausted
 	}
-	methodMap[call.method] = append(methodMap[call.method], call)
+	m[key] = append(m[key], call)
 }
 
 // Remove removes an expected call.
 func (cs callSet) Remove(call *Call) {
-	methodMap, ok := cs[call.receiver]
-	if !ok {
-		return
-	}
-	sl := methodMap[call.method]
-	for i, c := range sl {
+	key := callSetKey{call.receiver, call.method}
+	calls := cs.expected[key]
+	for i, c := range calls {
 		if c == call {
 			// maintain order for remaining calls
-			methodMap[call.method] = append(sl[:i], sl[i+1:]...)
+			cs.expected[key] = append(calls[:i], calls[i+1:]...)
+			cs.exhausted[key] = append(cs.exhausted[key], call)
 			break
 		}
 	}
@@ -52,26 +64,12 @@ func (cs callSet) Remove(call *Call) {
 
 // FindMatch searches for a matching call. Returns error with explanation message if no call matched.
 func (cs callSet) FindMatch(receiver interface{}, method string, args []interface{}) (*Call, error) {
-	methodMap, ok := cs[receiver]
-	if !ok {
-		return nil, errors.New("there are no expected method calls for that receiver")
-	}
-	calls, ok := methodMap[method]
-	if !ok {
-		return nil, fmt.Errorf("there are no expected calls of the method: %s for that receiver", method)
-	}
+	key := callSetKey{receiver, method}
 
-	// Search through the unordered set of calls expected on a method on a
-	// receiver.
+	// Search through the expected calls.
+	expected := cs.expected[key]
 	var callsErrors bytes.Buffer
-	for _, call := range calls {
-		// A call should not normally still be here if exhausted,
-		// but it can happen if, for instance, .Times(0) was used.
-		// Pretend the call doesn't match.
-		if call.exhausted() {
-			callsErrors.WriteString("\nThe call was exhausted.")
-			continue
-		}
+	for _, call := range expected {
 		err := call.matches(args)
 		if err != nil {
 			fmt.Fprintf(&callsErrors, "\n%v", err)
@@ -80,5 +78,31 @@ func (cs callSet) FindMatch(receiver interface{}, method string, args []interfac
 		}
 	}
 
+	// If we haven't found a match then search through the exhausted calls so we
+	// get useful error messages.
+	exhausted := cs.exhausted[key]
+	for _, call := range exhausted {
+		if err := call.matches(args); err != nil {
+			fmt.Fprintf(&callsErrors, "\n%v", err)
+		}
+	}
+
+	if len(expected)+len(exhausted) == 0 {
+		fmt.Fprintf(&callsErrors, "there are no expected calls of the method %q for that receiver", method)
+	}
+
 	return nil, fmt.Errorf(callsErrors.String())
+}
+
+// Failures returns the calls that are not satisfied.
+func (cs callSet) Failures() []*Call {
+	failures := make([]*Call, 0, len(cs.expected))
+	for _, calls := range cs.expected {
+		for _, call := range calls {
+			if !call.satisfied() {
+				failures = append(failures, call)
+			}
+		}
+	}
+	return failures
 }

--- a/gomock/callset_test.go
+++ b/gomock/callset_test.go
@@ -14,40 +14,45 @@
 
 package gomock
 
-import "testing"
+import (
+	"reflect"
+	"testing"
+)
+
+type receiverType struct{}
+
+func (receiverType) Func() {}
 
 func TestCallSetAdd(t *testing.T) {
-	methodVal := "TestMethod"
-	var receiverVal interface{} = "TestReceiver"
-	cs := make(callSet)
+	method := "TestMethod"
+	var receiver interface{} = "TestReceiver"
+	cs := newCallSet()
 
 	numCalls := 10
 	for i := 0; i < numCalls; i++ {
-		cs.Add(&Call{receiver: receiverVal, method: methodVal})
+		cs.Add(newCall(t, receiver, method, reflect.TypeOf(receiverType{}.Func)))
 	}
 
-	if len(cs) != 1 {
-		t.Errorf("expected only one reciever in callSet")
+	call, err := cs.FindMatch(receiver, method, []interface{}{})
+	if err != nil {
+		t.Fatalf("FindMatch: %v", err)
 	}
-	if numActualMethods := len(cs[receiverVal]); numActualMethods != 1 {
-		t.Errorf("expected only method on the reciever in callSet, found %d", numActualMethods)
-	}
-	if numActualCalls := len(cs[receiverVal][methodVal]); numActualCalls != numCalls {
-		t.Errorf("expected all %d calls in callSet, found %d", numCalls, numActualCalls)
+	if call == nil {
+		t.Fatalf("FindMatch: Got nil, want non-nil *Call")
 	}
 }
 
 func TestCallSetRemove(t *testing.T) {
-	methodVal := "TestMethod"
-	var receiverVal interface{} = "TestReceiver"
+	method := "TestMethod"
+	var receiver interface{} = "TestReceiver"
 
-	cs := make(callSet)
+	cs := newCallSet()
 	ourCalls := []*Call{}
 
 	numCalls := 10
 	for i := 0; i < numCalls; i++ {
 		// NOTE: abuse the `numCalls` value to convey initial ordering of mocked calls
-		generatedCall := &Call{receiver: receiverVal, method: methodVal, numCalls: i}
+		generatedCall := &Call{receiver: receiver, method: method, numCalls: i}
 		cs.Add(generatedCall)
 		ourCalls = append(ourCalls, generatedCall)
 	}
@@ -65,7 +70,7 @@ func TestCallSetRemove(t *testing.T) {
 	}
 
 	for _, c := range ourCalls {
-		validateOrder(cs[receiverVal][methodVal])
+		validateOrder(cs.expected[callSetKey{receiver, method}])
 		cs.Remove(c)
 	}
 }

--- a/gomock/controller.go
+++ b/gomock/controller.go
@@ -76,13 +76,13 @@ type TestReporter interface {
 type Controller struct {
 	mu            sync.Mutex
 	t             TestReporter
-	expectedCalls callSet
+	expectedCalls *callSet
 }
 
 func NewController(t TestReporter) *Controller {
 	return &Controller{
 		t:             t,
-		expectedCalls: make(callSet),
+		expectedCalls: newCallSet(),
 	}
 }
 
@@ -121,27 +121,16 @@ func (ctrl *Controller) RecordCall(receiver interface{}, method string, args ...
 }
 
 func (ctrl *Controller) RecordCallWithMethodType(receiver interface{}, method string, methodType reflect.Type, args ...interface{}) *Call {
-	// TODO: check arity, types.
-	margs := make([]Matcher, len(args))
-	for i, arg := range args {
-		if m, ok := arg.(Matcher); ok {
-			margs[i] = m
-		} else if arg == nil {
-			// Handle nil specially so that passing a nil interface value
-			// will match the typed nils of concrete args.
-			margs[i] = Nil()
-		} else {
-			margs[i] = Eq(arg)
-		}
+	if h, ok := ctrl.t.(testHelper); ok {
+		h.Helper()
 	}
+
+	call := newCall(ctrl.t, receiver, method, methodType, args...)
 
 	ctrl.mu.Lock()
 	defer ctrl.mu.Unlock()
-
-	origin := callerInfo(2)
-	call := &Call{t: ctrl.t, receiver: receiver, method: method, methodType: methodType, args: margs, origin: origin, minCalls: 1, maxCalls: 1}
-
 	ctrl.expectedCalls.Add(call)
+
 	return call
 }
 
@@ -200,18 +189,11 @@ func (ctrl *Controller) Finish() {
 	}
 
 	// Check that all remaining expected calls are satisfied.
-	failures := false
-	for _, methodMap := range ctrl.expectedCalls {
-		for _, calls := range methodMap {
-			for _, call := range calls {
-				if !call.satisfied() {
-					ctrl.t.Errorf("missing call(s) to %v", call)
-					failures = true
-				}
-			}
-		}
+	failures := ctrl.expectedCalls.Failures()
+	for _, call := range failures {
+		ctrl.t.Errorf("missing call(s) to %v", call)
 	}
-	if failures {
+	if len(failures) != 0 {
 		ctrl.t.Fatalf("aborting test due to missing call(s)")
 	}
 }

--- a/gomock/controller_test.go
+++ b/gomock/controller_test.go
@@ -169,7 +169,7 @@ func TestNoRecordedCallsForAReceiver(t *testing.T) {
 
 	reporter.assertFatal(func() {
 		ctrl.Call(subject, "NotRecordedMethod", "argument")
-	}, "Unexpected call to", "there are no expected method calls for that receiver")
+	}, "Unexpected call to", "there are no expected calls of the method \"NotRecordedMethod\" for that receiver")
 	ctrl.Finish()
 }
 
@@ -180,7 +180,7 @@ func TestNoRecordedMatchingMethodNameForAReceiver(t *testing.T) {
 	ctrl.RecordCall(subject, "FooMethod", "argument")
 	reporter.assertFatal(func() {
 		ctrl.Call(subject, "NotRecordedMethod", "argument")
-	}, "Unexpected call to", "there are no expected calls of the method: NotRecordedMethod for that receiver")
+	}, "Unexpected call to", "there are no expected calls of the method \"NotRecordedMethod\" for that receiver")
 	reporter.assertFatal(func() {
 		// The expected call wasn't made.
 		ctrl.Finish()

